SoundWire: revisit interrupt and lcount handling

plbossart opened the pull request:
The code in drivers/soundwire/intel_init.c is hardware-dependent and the code does not apply to new generations starting with MeteorLake. Refactor and clean-up the code to make this intel_init.c hardware-agnostic and move all hardware-dependencies in the SOF driver using chip descriptors.

Tested on TGL SKU 0A32

RanderWang commented on sound/soc/sof/intel/hda.c:
```diff
@@ -244,10 +244,45 @@ static int hda_sdw_probe(struct snd_sof_dev *sdev)
+		return -EINVAL;
+	}
+
+	return caps;
```
caps is u32 but return type is int

plbossart replied:
changed, now returns 0 on success and -EINVAL on errors.

bardliao commented on sound/soc/sof/intel/hda.c:
```diff
@@ -257,6 +292,10 @@ int hda_sdw_startup(struct snd_sof_dev *sdev)
 	if (pdata->machine && !pdata->machine->mach_params.link_mask)
 		return 0;
 
+	ret = hda_sdw_read_lcount(sdev);
```
Where will the `ret` be used? And will it be better to rename `ret` to `lcount`?

bardliao replied:
> Where will the `ret` be used? And will it be better to rename `ret` to `lcount`?

Oh, I see, we only check it. Not really using it.

plbossart replied:
I should rename this 'check_sdw_lcount', we've not using the result just verifying that what ACPI reported is aligned with what the hardware supports. Good suggestion.

RanderWang reviewed (approved):
LGTM

plbossart commented:
Going back to draft, I am not happy with the PR, it needs more work to move the startup and interrupt enabling.

plbossart commented:
@bardliao @RanderWang do you mind re-looking at the update? I only fixed a couple of issues here and there.

RanderWang reviewed (dismissed):
LGTM

ranj063 commented on sound/soc/sof/intel/mtl.c:
```diff
@@ -568,6 +576,7 @@ static void mtl_ipc_dump(struct snd_sof_dev *sdev)
 
 static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 {
+	hda_sdw_int_enable(sdev, false);
```
is this correct @plbossart . We have mtl_enable_sdw_irq() but you're not using that?

plbossart replied:
it's the same in the end. for MTL, a call to hda_sdw_int_enable() will result in a call  to mtl_enable_sdw_irq()

````c
void hda_sdw_int_enable(struct snd_sof_dev *sdev, bool enable)
{
	const struct sof_intel_dsp_desc *chip;

	chip = get_chip_info(sdev->pdata);
	if (chip && chip->enable_sdw_irq)
		chip->enable_sdw_irq(sdev, enable);
		
mtl.c
		.enable_sdw_irq = mtl_enable_sdw_irq,
````

I preferred to use the higher-level function for consistency with other platforms.

ujfalusi replied:
In this particular case I would just call `mtl_enable_sdw_irq(sdev, false)` for now, I expect to see further down the line that these steps will move to a generic code where the callback actually make sense.

With mtl code calling mtl function from mtl.c via abstraction makes things harder to follow.

plbossart replied:
ok, if there are 2 reviewers who think it's confusing that's a clear sign I was in the wrong. Followed the suggestion and used mtl_enable_sdw_irq(sdev, false)  directly without any abstraction/indirection

ranj063 reviewed (dismissed):
only one comment but looks great otherwise

plbossart commented:
let's give @bardliao a chance to review.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -155,9 +155,33 @@ struct sdw_intel_ops sdw_callback = {
+	else
+		val &= ~HDA_DSP_REG_ADSPIC2_SNDW;
+
+	snd_sof_dsp_write(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC2, val);
```
nitpick: we have `snd_sof_dsp_update_bits()` to 'hide' the read and write.
```c
snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC2, HDA_DSP_REG_ADSPIC2_SNDW,
			enable ? HDA_DSP_REG_ADSPIC2_SNDW : 0);
```

plbossart replied:
yes, good suggestion, thanks @ujfalusi

ujfalusi commented on sound/soc/sof/intel/shim.h:
```diff
@@ -185,6 +185,7 @@ struct sof_intel_dsp_desc {
 	u32 d0i3_offset;
 	u32 quirks;
 	enum sof_intel_hw_ip_version hw_ip_version;
+	void (*enable_sdw_irq)(struct snd_sof_dev *sdev, bool enable);
```
@plbossart, is it really because of the different programming sequence or just merely that different registers needs to be programmed as with the check irq, where the we need to read different registers.
Referring to the commit message.

plbossart replied:
it's both a different programming sequence and registers @ujfalusi, I clarified the commit message.

ujfalusi commented on sound/soc/sof/intel/shim.h:
```diff
@@ -185,6 +185,7 @@ struct sof_intel_dsp_desc {
 	u32 d0i3_offset;
 	u32 quirks;
 	enum sof_intel_hw_ip_version hw_ip_version;
+	int (*read_sdw_lcount)(struct snd_sof_dev *sdev);
```
Right, so this commit message is clear on the reasons to introduce the callback abstraction (different registers on different platforms).

plbossart replied:
no change then

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -983,12 +983,13 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	hda_codec_probe_bus(sdev);
 
+err:
 	if (!HDA_IDISP_CODEC(bus->codec_mask))
```
The `bus->codec_mask` is determined within `hda_dsp_ctrl_init_chip()` as first thing in hda_init_caps(), if there were IDISP identified the patch will not do what the commit message states, the i915 power will be not turned off?

plbossart replied:
indeed, nice catch @ujfalusi 

I will change the test to 
````c
if (!HDA_IDISP_CODEC(bus->codec_mask) || ret < 0)
    hda_codec_i915_display_power(sdev, false);
````

plbossart replied:
nope, this doesn't work. I need to add a boolean and test it.

plbossart commented:
changes since last version (comments from @ranj063 and @ujfalusi )

````diff
diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index bc32636c2014..944f95bd1e92 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -165,14 +165,9 @@ void hda_common_enable_sdw_irq(struct snd_sof_dev *sdev, bool enable)
        if (!hdev->sdw)
                return;
 
-       val = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC2);
-
-       if (enable)
-               val |= HDA_DSP_REG_ADSPIC2_SNDW;
-       else
-               val &= ~HDA_DSP_REG_ADSPIC2_SNDW;
-
-       snd_sof_dsp_write(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC2, val);
+       snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC2,
+                               HDA_DSP_REG_ADSPIC2_SNDW,
+                               enable ? HDA_DSP_REG_ADSPIC2_SNDW : 0);
 }
 
 void hda_sdw_int_enable(struct snd_sof_dev *sdev, bool enable)
@@ -935,6 +930,7 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
        struct hdac_bus *bus = sof_to_bus(sdev);
        struct snd_sof_pdata *pdata = sdev->pdata;
        struct sof_intel_hda_dev *hdev = pdata->hw_pdata;
+       bool i915_off = false;
        u32 link_mask;
        int ret;
 
@@ -975,6 +971,7 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
        ret = hda_sdw_probe(sdev);
        if (ret < 0) {
                dev_err(sdev->dev, "SoundWire probe error: %d\n", ret);
+               i915_off = true;
                goto err;
        }
 
@@ -984,7 +981,7 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
        hda_codec_probe_bus(sdev);
 
 err:
-       if (!HDA_IDISP_CODEC(bus->codec_mask))
+       if (!HDA_IDISP_CODEC(bus->codec_mask) || i915_off)
                hda_codec_i915_display_power(sdev, false);
 
        hda_bus_ml_put_all(bus);
diff --git a/sound/soc/sof/intel/mtl.c b/sound/soc/sof/intel/mtl.c
index d20f04a19ad6..904ae42534e1 100644
--- a/sound/soc/sof/intel/mtl.c
+++ b/sound/soc/sof/intel/mtl.c
@@ -576,7 +576,7 @@ static void mtl_ipc_dump(struct snd_sof_dev *sdev)
 
 static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 {
-       hda_sdw_int_enable(sdev, false);
+       mtl_enable_sdw_irq(sdev, false);
        mtl_disable_ipc_interrupts(sdev);
        return mtl_enable_interrupts(sdev, false);
 }
````

bardliao commented on sound/soc/sof/intel/hda.c:
```diff
@@ -966,8 +967,9 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
-		dev_err(sdev->dev, "error: SoundWire probe error\n");
-		return ret;
+		dev_err(sdev->dev, "SoundWire probe error: %d\n", ret);
+		i915_off = true;
```
Why `i915_off = true` is only set here? When will i915 be turned on? Do we need to set `i915_off = true` in every `return` and `goto` above?

plbossart replied:
dropped

ujfalusi commented on sound/soc/sof/intel/hda.h:
```diff
@@ -294,6 +294,7 @@
 #define HDA_DSP_REG_ADSPIS2		(HDA_DSP_GEN_BASE + 0x14)
 
+#define HDA_DSP_REG_ADSPIC2_SNDW	BIT(5)
 #define HDA_DSP_REG_ADSPIS2_SNDW	BIT(5)
```
I had to pause on this first. I though it is adding the same define again, then noticed the ADSPI**C**2 vs ADSPI**S**2.

plbossart replied:
yes, same bit in different registers

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -977,12 +979,13 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
-	if (!HDA_IDISP_CODEC(bus->codec_mask))
+err:
+	if (!HDA_IDISP_CODEC(bus->codec_mask) || i915_off)
 		hda_codec_i915_display_power(sdev, false);
```
On other note: how this supposed to working in the original case??
```c
void hda_codec_i915_display_power(struct snd_sof_dev *sdev, bool enable)
{
	struct hdac_bus *bus = sof_to_bus(sdev);

	if (IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC_DEBUG_SUPPORT) &&
	    sof_debug_check_flag(SOF_DBG_FORCE_NOCODEC))
		return;

	if (HDA_IDISP_CODEC(bus->codec_mask)) {
		dev_dbg(bus->dev, "Turning i915 HDAC power %d\n", enable);
		snd_hdac_display_power(bus, HDA_CODEC_IDX_CONTROLLER, enable);
	}
}
EXPORT_SYMBOL_NS_GPL(hda_codec_i915_display_power, SND_SOC_SOF_HDA_AUDIO_CODEC_I915);
```
If we only call `hda_codec_i915_display_power()` in case `!HDA_IDISP_CODEC(bus->codec_mask)` then the whole thing is a NOP?

The power initially turned on via hda_init()->hda_codec_i915_init()->snd_hdac_display_power() directly and if the hda_iit_caps() fails it is going to be turned off via the error handling: hda_dsp_probe()->hda_codec_i915_exit()

plbossart replied:
actually there are two problems with the code. First this:

````c
	/* scan SoundWire capabilities exposed by DSDT */
	ret = hda_sdw_acpi_scan(sdev);
	if (ret < 0) {
		dev_dbg(sdev->dev, "skipping SoundWire, not detected with ACPI scan\n");
		goto skip_soundwire;
	}
````

if ret < 0 then we would stop the probe. That's not the intended design and the idea with 'skip_soundwire'.

On error cases we do turn off the i915 as explained by @ujfalusi. The i915 power management is a bit misleading in the power is turned on/off in multiple places. It's unconditionally turned on/off in hda_codev_i915_init/exit and conditionally everyone else. that's fine, but @ujfalusi has a point that this is a no-op:

````
if (!HDA_IDISP_CODEC(bus->codec_mask))
		hda_codec_i915_display_power(sdev, false);
````

this code was added by @kv2019i in 0c75419a94a20b2bf791825a665d8b64cf7c560d, but our very own @kv2019i added an additional filter in 816938b272b0ac0203e25ce50483bd284ea4a2db which now makes the initial code a no-op.

One of the two changes is correct, which one @kv2019i ?

plbossart replied:
I'll drop this patch and added https://github.com/thesofproject/linux/issues/3993 to track the i915 thread.

plbossart commented:
dropped "ASoC: SOF: Intel: hda: improve error handling on SoundWire probe", no other change.
